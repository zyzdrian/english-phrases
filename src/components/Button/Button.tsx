import React from 'react';
import styled from 'styled-components';

interface IStyledButton {
    variant: 'primary' | 'secondary';
}

const StyledButton = styled.button<IStyledButton>`
flex-grow: 1;
flex-basis: 0;
background-color: ${props => props.theme[props.variant]};
box-shadow: ${props => props.theme.shadow};
border: none;
color: white;
border-radius: 16px;
cursor: pointer;
height: 56px;
font-weight: bold;
line-height: 1.36;
letter-spacing: 0.28px;

&:disabled {
  background-color: gray;
  color: white;
  cursor: default;
}

&:focus {
  outline: none;
}
`;

export const Button: React.FC<IButton> = ({ children, variant = 'primary', ...rest }) => (
    <StyledButton variant={variant} {...rest}>{children}</StyledButton>
);

export interface IButton {
    variant?: 'primary' | 'secondary';
    onClick?: () => void;
    disabled?: boolean;
}
