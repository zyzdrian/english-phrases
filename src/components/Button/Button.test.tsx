import React from 'react';
import { Button } from './Button';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { renderTheme, lightTheme } from '../../styledComponents';

describe('<Button /> tests', () => {
    it('renders without props', () => {
        const { container, getByRole } = renderTheme(<Button />);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByRole('button')).toHaveTextContent('');
        expect(getByRole('button')).not.toHaveAttribute('disabled');
        expect(getByRole('button')).toHaveStyle(`background-color: ${lightTheme.primary}`);
    });

    it('renders with children', () => {
        const text = 'Test text';

        const { container, getByRole } = render(<Button>{text}</Button>);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByRole('button')).toHaveTextContent(text);
    });

    it('renders with prop disabled', () => {
        const { container, getByRole } = render(<Button disabled />);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByRole('button')).toHaveAttribute('disabled');
    });

    it('renders as secondary', () => {
        const { container, getByRole } = renderTheme(<Button variant='secondary' />);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByRole('button')).toHaveStyle(`background-color: ${lightTheme.secondary}`);
    });

    it('renders with prop disabled', () => {
        const onClick = jest.fn();

        const { container, getByRole } = render(<Button onClick={onClick} />);
        fireEvent.click(getByRole('button'));

        expect(container.firstChild).toMatchSnapshot();
        expect(onClick).toBeCalled();
    });
});
