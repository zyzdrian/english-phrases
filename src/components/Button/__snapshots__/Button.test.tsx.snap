// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> tests renders as secondary 1`] = `
<button
  class="sc-bdVaJa gkfBsR"
/>
`;

exports[`<Button /> tests renders with children 1`] = `
<button
  class="sc-bdVaJa fnRXyy"
>
  Test text
</button>
`;

exports[`<Button /> tests renders with prop disabled 1`] = `
<button
  class="sc-bdVaJa fnRXyy"
  disabled=""
/>
`;

exports[`<Button /> tests renders with prop disabled 2`] = `
<button
  class="sc-bdVaJa fnRXyy"
/>
`;

exports[`<Button /> tests renders without props 1`] = `
<button
  class="sc-bdVaJa iVphkJ"
/>
`;
